Sources/SwiftDocC/Converter/RenderNode+Coding.swift

```diff
@@ -112,13 +112,13 @@ public enum RenderJSONEncoder {
         emitVariantOverrides: Bool = true
     ) -> JSONEncoder {
         let encoder = JSONEncoder()
-        
+
+        if #available(macOS 10.13, iOS 11.0, watchOS 4.0, tvOS 11.0, *) {
+          encoder.outputFormatting.insert(.sortedKeys)
```

Comment:
Moving this outside of the if-statement is a change in behavior that possibly should have some discussion about it. 

The only concern would be if there's a performance impact to sorting the render JSON in cases where the output doesn't need to be sorted.

There are some utilities in [bin/make-test-bundle](https://github.com/apple/swift-docc/tree/main/bin/make-test-bundle) and [bin/benchmark.swift](https://github.com/apple/swift-docc/blob/main/bin/benchmark.swift#L148-L151) that can be used to generate example projects of a few different sizes and gather performance measurements for building documentation. 

Together these can be used to compare these changes to the "main" branch to see if sorting the JSON keys in the output has any measurable performance impact.

Reply:
Excellent benchmarking tools! Here are the results:

## Benchmark Results

### sizeFactor 10

<details><summary>main (baseline)</summary>

```json
{
  "date" : "Feb 2, 2022 at 19:05:53",
  "metrics" : [
    {
      "result" : 4762,
      "identifier" : "duration-bundle-registration",
      "displayName" : "Duration for 'bundle-registration' (msec)"
    },
    {
      "result" : 8092,
      "identifier" : "duration-convert-action",
      "displayName" : "Duration for 'convert-action' (msec)"
    },
    {
      "result" : "edce22bce4f8de475a5e58a2e45dcdc0",
      "identifier" : "topic-graph-hash",
      "displayName" : "Topic Graph Checksum"
    },
    {
      "result" : "c8f03d4e81da8e3b6d7fb7f63c771147",
      "identifier" : "topic-anchor-hash",
      "displayName" : "Topic Anchor Checksum"
    },
    {
      "result" : 456564736,
      "identifier" : "peak-memory",
      "displayName" : "Peak memory footprint (bytes)"
    },
    {
      "result" : 120441674,
      "identifier" : "output-size",
      "displayName" : "Compiled output size (bytes)"
    }
  ],
  "arguments" : [
    "convert",
    "\/var\/folders\/px\/9trts7nj3zv_smktjpzxqqp80000gp\/T\/45D70E9D-36C2-4988-8275-834AB3264E3A.docc"
  ],
  "platform" : "macOS"
}
```
</details>

<details><summary>bcr/2202/sorted-keys</summary>

```json
{
  "arguments" : [
    "convert",
    "\/var\/folders\/px\/9trts7nj3zv_smktjpzxqqp80000gp\/T\/4EBB14F5-CCE6-442B-9E23-44FF1C49388D.docc"
  ],
  "date" : "Feb 2, 2022 at 19:36:00",
  "metrics" : [
    {
      "identifier" : "duration-bundle-registration",
      "result" : 4568,
      "displayName" : "Duration for 'bundle-registration' (msec)"
    },
    {
      "identifier" : "duration-convert-action",
      "result" : 8278,
      "displayName" : "Duration for 'convert-action' (msec)"
    },
    {
      "identifier" : "topic-graph-hash",
      "result" : "edce22bce4f8de475a5e58a2e45dcdc0",
      "displayName" : "Topic Graph Checksum"
    },
    {
      "identifier" : "topic-anchor-hash",
      "result" : "c8f03d4e81da8e3b6d7fb7f63c771147",
      "displayName" : "Topic Anchor Checksum"
    },
    {
      "identifier" : "peak-memory",
      "result" : 452464640,
      "displayName" : "Peak memory footprint (bytes)"
    },
    {
      "identifier" : "output-size",
      "result" : 120441674,
      "displayName" : "Compiled output size (bytes)"
    }
  ],
  "platform" : "macOS"
}
```
</details>

#### diff

```
+-----------------------------------------------------------------------------------------------------+
| Metric                                   | Change     | Before               | After                |
+-----------------------------------------------------------------------------------------------------+
| Compiled output size (MB)                | no change  | 114.86               | 114.86               |
| Duration for 'bundle-registration' (sec) | -4.07%     | 4.76                 | 4.57                 |
| Duration for 'convert-action' (sec)      | +2.30%     | 8.09                 | 8.28                 |
| Peak memory footprint (MB)               | -0.90%     | 435.41               | 431.50               |
| Topic Anchor Checksum                    | no change  | c8f03d4e81da8e3b6d7f | c8f03d4e81da8e3b6d7f |
| Topic Graph Checksum                     | no change  | edce22bce4f8de475a5e | edce22bce4f8de475a5e |
+-----------------------------------------------------------------------------------------------------+
```

### sizeFactor 100

<details><summary>main (baseline)</summary>

```json
{
  "arguments" : [
    "convert",
    "\/var\/folders\/px\/9trts7nj3zv_smktjpzxqqp80000gp\/T\/93BDD94E-E672-44C2-9CCB-B2B349129D63.docc"
  ],
  "metrics" : [
    {
      "result" : 54852,
      "identifier" : "duration-bundle-registration",
      "displayName" : "Duration for 'bundle-registration' (msec)"
    },
    {
      "result" : 96365,
      "identifier" : "duration-convert-action",
      "displayName" : "Duration for 'convert-action' (msec)"
    },
    {
      "result" : "f6fed1b864f9b26c0f74913849b153b7",
      "identifier" : "topic-graph-hash",
      "displayName" : "Topic Graph Checksum"
    },
    {
      "result" : "9f19266e9b0ef156230e9c3495a28dec",
      "identifier" : "topic-anchor-hash",
      "displayName" : "Topic Anchor Checksum"
    },
    {
      "result" : 4287963136,
      "identifier" : "peak-memory",
      "displayName" : "Peak memory footprint (bytes)"
    },
    {
      "result" : 1217682269,
      "identifier" : "output-size",
      "displayName" : "Compiled output size (bytes)"
    }
  ],
  "platform" : "macOS",
  "date" : "Feb 2, 2022 at 19:26:59"
}
```
</details>

<details><summary>bcr/2202/sorted-keys</summary>

```json
{
  "platform" : "macOS",
  "date" : "Feb 2, 2022 at 19:57:16",
  "arguments" : [
    "convert",
    "\/var\/folders\/px\/9trts7nj3zv_smktjpzxqqp80000gp\/T\/277750C2-24B9-41CC-BC8D-58DD781638E7.docc"
  ],
  "metrics" : [
    {
      "identifier" : "duration-bundle-registration",
      "result" : 52019,
      "displayName" : "Duration for 'bundle-registration' (msec)"
    },
    {
      "identifier" : "duration-convert-action",
      "result" : 93664,
      "displayName" : "Duration for 'convert-action' (msec)"
    },
    {
      "identifier" : "topic-graph-hash",
      "result" : "f6fed1b864f9b26c0f74913849b153b7",
      "displayName" : "Topic Graph Checksum"
    },
    {
      "identifier" : "topic-anchor-hash",
      "result" : "9f19266e9b0ef156230e9c3495a28dec",
      "displayName" : "Topic Anchor Checksum"
    },
    {
      "identifier" : "peak-memory",
      "result" : 4322598912,
      "displayName" : "Peak memory footprint (bytes)"
    },
    {
      "identifier" : "output-size",
      "result" : 1217682269,
      "displayName" : "Compiled output size (bytes)"
    }
  ]
}
```
</details>

#### diff

```
+-----------------------------------------------------------------------------------------------------+
| Metric                                   | Change     | Before               | After                |
+-----------------------------------------------------------------------------------------------------+
| Compiled output size (MB)                | no change  | 1161.27              | 1161.27              |
| Duration for 'bundle-registration' (sec) | -5.16%     | 54.85                | 52.02                |
| Duration for 'convert-action' (sec)      | -2.80%     | 96.36                | 93.66                |
| Peak memory footprint (MB)               | +0.81%     | 4089.32              | 4122.35              |
| Topic Anchor Checksum                    | no change  | 9f19266e9b0ef156230e | 9f19266e9b0ef156230e |
| Topic Graph Checksum                     | no change  | f6fed1b864f9b26c0f74 | f6fed1b864f9b26c0f74 |
+-----------------------------------------------------------------------------------------------------+
```

Reply:
Apologies for the delayed response. Since this change only applies to the rendering portion of the convert action I'm surprised to see such a large difference in the bundle registration time, which is the same code in both runs. This makes me wonder if the machine was under significantly different load or if there's something else that's skewing the results. 

Seeing that the total convert action time didn't follow the bundle registration change also makes me wonder if there is a impact to the rendering time.

I ran two benchmarks with a size factor 10 and got the following results:
```
+-----------------------------------------------------------------------------------------------------+
| Metric                                   | Change     | Before               | After                |
+-----------------------------------------------------------------------------------------------------+
| Compiled output size (MB)                | no change  | 106.71               | 106.71               |
| Duration for 'bundle-registration' (sec) | +0.28%     | 4.23                 | 4.24                 |
| Duration for 'convert-action' (sec)      | +3.91%     | 5.90                 | 6.13                 |
| Peak memory footprint (MB)               | +0.55%     | 413.14               | 415.42               |
| Topic Anchor Checksum                    | no change  | c8f03d4e81da8e3b6d7f | c8f03d4e81da8e3b6d7f |
| Topic Graph Checksum                     | no change  | edce22bce4f8de475a5e | edce22bce4f8de475a5e |
+-----------------------------------------------------------------------------------------------------+
```
and
```
+-----------------------------------------------------------------------------------------------------+
| Metric                                   | Change     | Before               | After                |
+-----------------------------------------------------------------------------------------------------+
| Compiled output size (MB)                | no change  | 106.71               | 106.71               |
| Duration for 'bundle-registration' (sec) | +0.07%     | 4.25                 | 4.26                 |
| Duration for 'convert-action' (sec)      | +1.86%     | 5.95                 | 6.06                 |
| Peak memory footprint (MB)               | +0.33%     | 415.07               | 416.46               |
| Topic Anchor Checksum                    | no change  | c8f03d4e81da8e3b6d7f | c8f03d4e81da8e3b6d7f |
| Topic Graph Checksum                     | no change  | edce22bce4f8de475a5e | edce22bce4f8de475a5e |
+-----------------------------------------------------------------------------------------------------+
```

Only five samples per run (which is what I used in my measurements) isn't a lot of data to analyze but it does give some indication that there could be a small performance impact. 

This makes me think that the next step is to discuss how important the sorted output is. Do we expect developers who need sorted output to enable pretty printing or do we expect the output to always be sorted. For example, do teams that commit documentation archives to source control also care about the visual diffability of pretty printed output or only that the compact JSON output doesn't change between identical builds.

What do you think @franklinsch

Reply:
As a developer who'd love this capability to be included, my primary goal isn't for viewing the content in sorted order, but to have the output be **deterministic** - in particular when rendering docs. Right now two renders of the _same_ content will generate drastically different content - and when I'm stashing said content in Github pages, it would be nice if the only updates corresponding to updates in the underlying code or catalog. 

My ideal end result here is:
- run the convert, commit the changes (for GH pages)
- run the convert again (with NO changes) - and see no reported changes from git.

Today that's not the case. It's not critical mind you, but it would be an extremely nice quality of life improvement for those of us hosting docs on Github pages.

Reply:
Hi @heckj,

I agree that having the output be deterministic **by default** would be a significant quality of life improvement here. However, there is an environment variable you can set when converting documentation to achieve this in the mean time.

```shell
# Pretty print DocC JSON output so that it can be consistently diffed between commits
export DOCC_JSON_PRETTYPRINT="YES"
```

The `swift-docc-plugin` repo has an `update-gh-pages-documentation-site` script that _might_ be helpful here and does set this variable for the reasons you outlined above: https://github.com/apple/swift-docc-plugin/blob/main/bin/update-gh-pages-documentation-site#L36.

Again, completely agree that having this by default would be great, just wanted to provide this to unblock you in the meantime.

Reply:
@ethan-kusters setting that environment variable still doesn't output deterministic JSON because of the inadequate API availability  check. I could move the (updated) availability check back into the prettyPrint check if that's the preference here. 

For my purpose it doesn't matter if the json is pretty-printed or not, although the pretty-printing is intended for debugging and the deterministic output is unrelated to debugging, so coupling them together seems incorrect.

Perhaps another environment variable (or CLI flag) for deterministic output, documented as part of the public API, so that the feature is opt-in?

Reply:
Just catching up on this - I also share concerns regarding performance with sorting keys by default. The JSON that DocC produces is optimised by machine readability and size first-and-foremost. The `DOCC_JSON_PRETTYPRINT` environment variable can be used in development environments to help debug issues. It would be nice to see `DOCC_JSON_PRETTYPRINT` augmented to also sort keys, as it would help in the development workflow.

It seems that the purpose of this change from your standpoint is **not** for debugging purposes though, and instead you'd like to compare documentation archives so that they're only committed if there are changes, is that right? Since the JSON spec doesn't enforce ordering of keys in dictionaries, have you considered using separate tooling (other than `diff`) that compares JSON semantically (i.e., agnostic towards ordering)? You could also consider linting the JSON files before comparing them in your tooling, for example via `jq --sort-keys . render_json.json > render_json.json`.

Reply:
Hey @franklinsch , sorry for the delayed response.

> I also share concerns regarding performance with sorting keys by default.

I'm aligned here; it doesn't make sense to introduce performance degradation to all users to solve a problem for a subset of users. I can update this PR to make the feature opt-in. What's the preferred approach? Another environment flag? Alternatively, I can lump it in with `DOCC_JSON_PRETTYPRINT`. Even though the json sorting isn't for debugging, anybody who is sorting keys is less concerned about performance and, at least for me, the size of the json files doesn't matter. Also, I can argue that sorting the JSON keys is a useful debugging tool as well.

> It seems that the purpose of this change from your standpoint is not for debugging purposes though, and instead you'd like to compare documentation archives so that they're only committed if there are changes, is that right?

Correct

> Since the JSON spec doesn't enforce ordering of keys in dictionaries

But `JSONEncoder.OutputFormatting.sortedKeys` does and is provided natively, so I don't think the JSON spec is relevant. I have to sort it either way, and I expect other people will also want to.

> have you considered using separate tooling (other than diff) that compares JSON semantically (i.e., agnostic towards ordering)? You could also consider linting the JSON files before comparing them in your tooling

Yes, I could certainly do that, and everybody else who needs this feature could do that too. But if it's done here, that eliminates an extra tool needed to solve a problem that can be solved here.

Reply:
I think adding this behaviour to when `DOCC_JSON_PRETTYPRINT` is set would be a fantastic improvement indeed. Thanks @bricker!

Reply:
@franklinsch Done!

Reply:
@franklinsch Done! Thanks.